File: dataStructure/src/queue/circularQueue.c
#include "ds_queue.h"

/** Initialize your data structure here. Set the size of the queue to be k. */
MyCircularQueue* myCircularQueueCreate(int k){
	MyCircularQueue* mcq;
    
    mcq = (MyCircularQueue*)malloc( sizeof(MyCircularQueue) );

    printf("[%p]\n", mcq );
	
    mcq->queue = (int*)malloc( sizeof(int) * k + 1 );
    mcq->head = mcq->queue;
    mcq->tail = mcq->queue + k - 1;
    mcq->front = 0;
    mcq->last = -1;
    mcq->size = k;
    mcq->currentSize = 0;

    printf("[%p]\n", mcq );

    return mcq;
}

/** Insert an element into the circular queue. Return true if the operation is successful. */
bool myCircularQueueEnQueue(MyCircularQueue* obj, int value) {
    if( obj->currentSize >= obj->size ){
        return false;
    }

    obj->currentSize++;
    /*
    obj->tail++;
    if( obj->tail - obj->queue == obj->size ){
        obj->tail = obj->queue;
    }

    *(obj->tail) = value;
    */
    obj->last++;
    *(obj->queue + obj->last%obj->size) = value;

    return true;
}

/** Delete an element from the circular queue. Return true if the operation is successful. */
bool myCircularQueueDeQueue(MyCircularQueue* obj) {
    if( obj->currentSize <= 0 ){
        obj->head = obj->queue;
        obj->tail = obj->queue + obj->size - 1;
        obj->last = -1;
        obj->front = 0;
        return false;
    }

    obj->currentSize--;
    /*
    *(obj->head) = -1;
    obj->head++;
    if( obj->head - obj->queue == obj->size  ){
        obj->head = obj->queue;
    }
    */
    *(obj->queue + obj->front % obj->size ) = -1;
    obj->front++;

    return true;
}

/** Get the front item from the queue. */
int myCircularQueueFront(MyCircularQueue* obj) {
    if( obj->currentSize > 0 ){
        return *(obj->queue + obj->front%obj->size );
//        return *(obj->head);
    }

    return -1;
}

/** Get the last item from the queue. */
int myCircularQueueRear(MyCircularQueue* obj) {
    if( obj->currentSize > 0 ){
        return *(obj->queue + obj->last%obj->size );
//        return *(obj->tail);
    }
  
    return -1;
}

/** Checks whether the circular queue is empty or not. */
bool myCircularQueueIsEmpty(MyCircularQueue* obj) {
    if( obj->currentSize == 0 ){
        return true;
    }
  
    return false;
}

/** Checks whether the circular queue is full or not. */
bool myCircularQueueIsFull(MyCircularQueue* obj) {
    if( obj->currentSize == obj->size ){
        return true;
    }
  
    return false;
}

void myCircularQueueFree(MyCircularQueue* obj) {
    if( obj->queue != NULL ){
        free( obj->queue );
        obj->head = NULL;
        obj->tail = NULL;
        obj->queue= NULL;
    }

    if( obj != NULL ){
        free(obj);
        obj = NULL;
    }
}

void MyCircularQueueItera(MyCircularQueue* obj) {
    int* ptr = NULL;
    ptr = obj->queue;
    for(int i=0; i<obj->size; i++){
        printf( "[%d]\t", *ptr++);
    }
}

File: dataStructure/src/queue/circularQueue.py
# -*- coding:utf-8 -*-

class MyCircularQueue:

    def __init__(self, k: int):
        """
        Initialize your data structure here. Set the size of the queue to be k.
        """
        self.data = [0] * k;
        self.head = 0;
        self.tail = -1;
        self.size = k;
        self.currentSize = 0;

    def enQueue(self, value: int) -> bool:
        """
        Insert an element into the circular queue. Return true if the operation is successful.
        """
        if( self.currentSize >= self.size ):
            return False;

        self.currentSize += 1;
        self.tail += 1;
        self.tail = (self.tail)%self.size;
        self.data[self.tail] = value;
        return True;

    def deQueue(self) -> bool:
        """
        Delete an element from the circular queue. Return true if the operation is successful.
        """
        if( self.currentSize <= 0 ):
            self.head = 0;
            self.tail = -1;
            return False;

        self.currentSize -= 1;
        self.data[self.head%self.size] = -1;
        self.head += 1;
        return True;

    def Front(self) -> int:
        """
        Get the front item from the queue.
        """
        if( self.isEmpty ):
            return -1;
        
        return self.data[self.head%self.size];
        

    def Rear(self) -> int:
        """
        Get the last item from the queue.
        """
        if( self.isEmpty ):
            return -1;
        
        return self.data[self.tail%self.size];
        

    def isEmpty(self) -> bool:
        """
        Checks whether the circular queue is empty or not.
        """
        if( self.currentSize == 0 ):
            return True;
        return False;

    def isFull(self) -> bool:
        """
        Checks whether the circular queue is full or not.
        """
        if( self.currentSize == self.size ):
            return True;
        return False;
    
    def viewQueue(self):
        print(self.data);

# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()
def main():
    cq = MyCircularQueue(3);
    print(cq.head);
    print(cq.tail);
    cq.viewQueue();

    cq.enQueue(2);
    print(cq.head);
    print(cq.tail);
    cq.enQueue(3);
    cq.enQueue(4);
    print(cq.head);
    print(cq.tail);
    cq.enQueue(5);
    cq.viewQueue();

    cq.deQueue();
    cq.deQueue();
    cq.deQueue();
    cq.enQueue(6);
    print(cq.head);
    print(cq.tail);
    cq.viewQueue();
    
    cq.enQueue(7);
    cq.deQueue();
    cq.enQueue(8);
    cq.viewQueue();

    print(cq.head);
    print(cq.tail);SJPLBCHBKHIIQCLF

if __name__ == "__main__":
    main();